src/server/frontend/wayland/wayland_connector.cpp

```diff
@@ -1967,62 +1954,207 @@ class WlShellSurface : public wayland::ShellSurface
+            params.aux_rect_placement_gravity = mir_placement_gravity_southeast;
+            params.placement_hints = mir_placement_hints_slide_x;
+            params.aux_rect_placement_offset_x = 0;
+            params.aux_rect_placement_offset_y = 0;
+        }
     }
 
     void set_maximized(std::experimental::optional<struct wl_resource*> const& output) override
     {
-        mir::shell::SurfaceSpecification mods;
-        mods.state = mir_window_state_maximized;
-        if (output)
+        if (surface_id.as_value())
         {
-            // TODO{alan_g} mods.output_id = DisplayConfigurationOutputId_from(output)
+            mir::shell::SurfaceSpecification mods;
+            mods.state = mir_window_state_maximized;
+            if (output)
+            {
+                // TODO{alan_g} mods.output_id = DisplayConfigurationOutputId_from(output)
+            }
+            auto const session = session_for_client(client);
+            shell->modify_surface(session, surface_id, mods);
+        }
+        else
+        {
+            params.state = mir_window_state_maximized;
+            if (output)
+            {
+                // TODO{alan_g} params.output_id = DisplayConfigurationOutputId_from(output)
+            }
         }
-        auto const session = session_for_client(client);
-        shell->modify_surface(session, surface_id, mods);
     }
 
-    void set_title(std::string const& /*title*/) override
+    void set_title(std::string const& title) override
     {
+        if (surface_id.as_value())
+        {
+            shell::SurfaceSpecification new_spec;
+            new_spec.name = title;
+            auto const session = session_for_client(client);
+            shell->modify_surface(session, surface_id, new_spec);
+        }
+        else
+        {
+            params.name = title;
+        }
     }
 
     void set_class(std::string const& /*class_*/) override
     {
     }
+
 private:
+    void resize_handler(
+        WlSurface* mir_surface,
+        std::shared_ptr<SurfaceEventSink> const& sink,
+        WlSeat& seat,
+        Size initial_size)
+    {
+        auto const session = session_for_client(client);
+
+        if (surface_id.as_value())
+        {
+            auto const surface = get_surface_for_id(session, surface_id);
+            if (surface->size() == initial_size)
+                return;
+            sink->latest_resize(initial_size);
+            shell::SurfaceSpecification new_size_spec;
+            new_size_spec.width = initial_size.width;
+            new_size_spec.height = initial_size.height;
+            shell->modify_surface(session, surface_id, new_size_spec);
+            return;
+        }
+
+        params.size = initial_size;
+        params.content_id = mir_surface->stream_id;
+        surface_id = shell->create_surface(session, params, sink);
+        mir_surface->surface_id = surface_id;
+
+        {
+            // The shell isn't guaranteed to respect the requested size
```

Comment:
Any idea if this is a Wayland  policy too?

Reply:
No direct knowledge, but I don't see how else things can be done e.g. for a tiling WM.